textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -33,9 +33,138 @@
 	
 	GramBooleanQuery(QueryOp operator, int gramLength) {
 		this.operator = operator;
-		operandList = new ArrayList<String>();
-		subQueryList = new ArrayList<GramBooleanQuery>();
-		this.gramLength = gramLength;
+		operandSet = new HashSet<String>();
+		subQuerySet = new HashSet<GramBooleanQuery>();
+		GramBooleanQuery.gramLength = gramLength;
+	}
+	
+	
+	/**
+	 * This method takes a list of strings and adds them to the query tree. <br>
+	 * For example, if the list is {abcd, wxyz}, then: <br>
+	 * trigrams({abcd, wxyz}) = trigrams(abcd) OR trigrams(wxyz) <br>
+	 * OR operator is assumed for a list of strings. <br>
+	 * @param list, a list of strings to be added into query.
+	 */
+	void add(List<String> list) {
+		addOrNode(list);
+	}
+	
+	private void addOrNode(List<String> literalList) {
+		if (TranslatorUtils.minLenOfString(literalList) < 3) {
```

Comment:
If the length of a literal is smaller than 3 (gramLength), we can't represent it with trigram. So we won't add it to the query. I think this could be customised.